components/compositing/compositor.rs

```diff
@@ -501,6 +501,26 @@ impl<Window: WindowMethods + ?Sized> IOCompositor<Window> {
         self.shutdown_state = ShutdownState::FinishedShuttingDown;
     }
 
+    pub fn pause(&mut self) {
+        debug!("Pausing compositor");
+        if let Err(e) = self.webrender_surfman.unbind_native_surface_from_context() {
+            warn!("Unbinding native surface from context failed ({:?})", e);
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn resume(&mut self, native_widget: *mut c_void, coords: DeviceIntSize) {
+        debug!("Resuming compositor: {native_widget:?}");
+        // Reinitialize webrender surfman and bind to context
+        let connection = self.webrender_surfman.connection();
+        let native_widget = unsafe {
+            connection.create_native_widget_from_ptr(native_widget, coords.to_untyped())
+        };
+        if let Err(e) = self.webrender_surfman.bind_native_surface_to_context(native_widget) {
+            warn!("Binding native surface to context failed ({:?})", e);
+        }
+    }
+
```

Comment:
Hrm...so normally the embedder is responsible for binding the native widget to the context, right? Is there a way to rework this so that this always happens in the embedder instead of making this happen in WebRender. Perhaps we could add a method to unset `webrender_surfman` in the compositor and then reset it from the embedder?

Reply:
I'm not sure if I fully understand the suggestion. 

> instead of making this happen in WebRender.

Do you mean in WebRenderSurfman ?

Today, AFAICT, the embedder doesn't  bind the widget directly, but it simply does `WebrenderSurfman::create` during initialization and then holds on to that instance for the lifetime. Are you suggesting that we move away from that model and make embedder recreate `WebrenderSurfman` on demand? Today  `WebrenderSurfman::create` also creates other resources like handle to the device, opengl context and swapchain - are we fine with recreating all these other resources as well?

Also, today the compositor  holds on the same WebrenderSurfman instance that the embedder creates (The data is wrapped in Rc and WindowMethods trait allows libservo to create a new handle from an open window). If we reset webrender_surfman in embedder, wouldn't we also need to reset it from compositor?

Reply:
I see. Thanks for the explanation! I think this approach is okay for now! We probably need to take a holistic look at how all these surfaces are managed at some point in the future.